ntb_hw_switchtec.c

```diff
@@ -587,14 +587,22 @@ enum switchtec_msg {
 
 static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev);
 
+static int crosslink_setup_req_ids(struct switchtec_ntb *sndev,
+		struct ntb_ctrl_regs __iomem *mmio_ctrl);
+
 static void switchtec_ntb_link_status_update(struct switchtec_ntb *sndev)
 {
 	int link_sta;
 	int old = sndev->link_is_up;
 
 	link_sta = sndev->self_shared->link_sta;
 	if (link_sta) {
-		u64 peer = ioread64(&sndev->peer_shared->magic);
+		u64 peer;
+
+		if (!sndev->link_is_up && crosslink_is_enabled(sndev))
+			crosslink_setup_req_ids(sndev, sndev->mmio_xlink_peer_ctrl);
```

Comment:
This and the subsequent patches don't quite make sense to me. If the requester Ids are not setup then will reading link_sta even work? It's supper weird and confusing to be doing setup when checking if the link is up.

Why do the requester ids need to wait until now to be setup? It doesn't seem like they depend on anything in the peer anyway....

May be just leave the setup in the enable and set a boolean, such that if it's not enabled return early when checking the link status...

Reply:
Let me try to explain it with a setup like below.

![image](https://user-images.githubusercontent.com/30617003/129657245-8a850e69-5584-4377-94e5-b3e9a5f85fef.png)

- For an access from host A to host B, when the TLP is routed from crosslink part A to crosslink part B, crosslink part B will only accept it when there's proper requester IDs configured.
- The requester IDs of crosslink part B are the proxy IDs of crosslink part A.
- The requester IDs of crosslink part B must be configured by host A, when switch B is in place and the crosslink cable is well connected. 

With above facts and back to your questions:
- Reading local link_sta (= sndev->self_shared->link_sta) doesn't require a crosslink requester ID setup.
- Requester IDs setup for crosslink part B initiated by host A may fail if switch B is not ready or no cable connected between two switches. This is the reason that I pushed the requester ID setup to the last possible moment. For example, when loading driver on host A or enabling link on host A, switch B may not be in a good state to receive the requester ID setup.

Since host A is setting up requester IDs in switch B (crosslink partition), it may conflict with other setups initiated by host B on crosslink partition B (like LUT setup). So an approach of checking current NT partition locking state (leveraging firmware capability) was introduced to resolve the competition.  

Hope it explains the thing...

Reply:
Yeah, ok, that makes sense. I didn't notice the link_sta was the local copy.

Can we add some of this explanation to the commit message here?